File: get_introspection.js
function intuneGetToken() {

  const requestUrl = "http://neogeek.io:4000/graphql?";
  var options = {};
  var payload = JSON.stringify({
    "query": "\n    query IntrospectionQuery {\n      __schema {\n        \n        queryType { name }\n        mutationType { name }\n        subscriptionType { name }\n        types {\n          ...FullType\n        }\n        directives {\n          name\n          description\n          \n          locations\n          args {\n            ...InputValue\n          }\n        }\n      }\n    }\n\n    fragment FullType on __Type {\n      kind\n      name\n      description\n      \n      fields(includeDeprecated: true) {\n        name\n        description\n        args {\n          ...InputValue\n        }\n        type {\n          ...TypeRef\n        }\n        isDeprecated\n        deprecationReason\n      }\n      inputFields {\n        ...InputValue\n      }\n      interfaces {\n        ...TypeRef\n      }\n      enumValues(includeDeprecated: true) {\n        name\n        description\n        isDeprecated\n        deprecationReason\n      }\n      possibleTypes {\n        ...TypeRef\n      }\n    }\n\n    fragment InputValue on __InputValue {\n      name\n      description\n      type { ...TypeRef }\n      defaultValue\n      \n      \n    }\n\n    fragment TypeRef on __Type {\n      kind\n      name\n      ofType {\n        kind\n        name\n        ofType {\n          kind\n          name\n          ofType {\n            kind\n            name\n            ofType {\n              kind\n              name\n              ofType {\n                kind\n                name\n                ofType {\n                  kind\n                  name\n                  ofType {\n                    kind\n                    name\n                  }\n                }\n              }\n            }\n          }\n        }\n      }\n    }\n  ",
    
  });
  options.method = "post";
  options.contentType = "application/json";
  options.payload = payload;
  
  var response = 0;
  try {
    response = UrlFetchApp.fetch(requestUrl, options);
  } catch(e) {
    Logger.log(e);
  }

  var results = JSON.parse(response.getContentText());

  var inputObjectList = [];
  var objectList = [];
  var scalarList = [];
  var enumList = [];
  var queryList = [];
  var mutationList = [];

  var queryTypeName = results.data.__schema.queryType.name;
  var mutationTypeName = results.data.__schema.mutationType.name;

  var typeData = results.data.__schema.types;

  typeData.forEach(function (d, i) {
    if(d.kind == "INPUT_OBJECT"){
      let object = {};
      object.name = d.name;
      object.inputFields = d.inputFields;
    }else if(d.kind == "OBJECT"){
      if(d.name == queryTypeName){
        d.fields.forEach(function(d,i){
          let object = {};
          object.name = d.name;
          object.type = d.type;
          object.args = [];
          d.args.forEach(function(d,i){
            let arg = {};
            arg.name = d.name;
            object.args.push(arg);
          });
          queryList.push(object);
        })
      }else if(d.name == mutationTypeName){
        d.fields.forEach(function(d,i){
          let object = {};
          object.name = d.name;
          object.type = d.type;
          object.args = [];
          d.args.forEach(function(d,i){
            let arg = {};
            arg.name = d.name;
            object.args.push(arg);
          });
          mutationList.push(object);
        })
      }else{
        let object = {};
        d.fields.forEach(function(d,i){
          object.name = d.name;
          object.type = d.type.kind;
        })
        objectList.push(object);
      }
    }else if(d.kind == "SCALAR"){
      scalarList.push(d);
    }else if(d.kind == "ENUM"){
      enumList.push(d);
    }
  });

  return results;

}

File: test/get_introspection_.py
import requests, functools, json
import yaml
from pprint import pprint

def send_request():
    url = "http://neogeek.io:4000/graphql"


    body = {
        "query": """query IntrospectionQuery {
      __schema {

        queryType { name }
        mutationType { name }
        subscriptionType { name }
        types {
          ...FullType
        }
        directives {
          name
          description

          locations
          args {
            ...InputValue
          }
        }
      }
    }

    fragment FullType on __Type {
      kind
      name
      description

      fields(includeDeprecated: true) {
        name
        description
        args {
          ...InputValue
        }
        type {
          ...TypeRef
        }
        isDeprecated
        deprecationReason
      }
      inputFields {
        ...InputValue
      }
      interfaces {
        ...TypeRef
      }
      enumValues(includeDeprecated: true) {
        name
        description
        isDeprecated
        deprecationReason
      }
      possibleTypes {
        ...TypeRef
      }
    }

    fragment InputValue on __InputValue {
      name
      description
      type { ...TypeRef }
      defaultValue


    }

    fragment TypeRef on __Type {
      kind
      name
      ofType {
        kind
        name
        ofType {
          kind
          name
          ofType {
            kind
            name
            ofType {
              kind
              name
              ofType {
                kind
                name
                ofType {
                  kind
                  name
                  ofType {
                    kind
                    name
                  }
                }
              }
            }
          }
        }
      }
    }

        """
    }


    x = requests.post(
        url=url,
        json=body
    )

    return json.loads(x.text)

def recursive_oftype(data):
  if(data["ofType"] == None):
    return {"kind" : data["kind"], "name" : data["name"]}
  else:
    return {"kind" : data["kind"], "name" : data["name"], "ofType" : recursive_oftype(data["ofType"])}


def parse(inspection_json):
    input_object_list = []
    object_list = []
    scalar_list = []
    enum_list = []
    query_list = []
    mutation_list = []

    query_type_name = inspection_json["data"]["__schema"]["queryType"]["name"]
    mutation_type_name = inspection_json["data"]["__schema"]["mutationType"]["name"]
    type_data = inspection_json["data"]["__schema"]["types"]

    for d in type_data:
        if d["kind"] == "INPUT_OBJECT":
            object = {}
            object["name"] = d["name"]
            object["inputFields"] = d["inputFields"]
            input_object_list.append(object)
        elif d["kind"] == "OBJECT":
            if d["name"] == query_type_name:
                for f in d["fields"]:
                    object = {}
                    object["name"] = f["name"]
                    types = {}
                    types = recursive_oftype(f["type"])

                    object["produces"] = types

                    object["consumes"] = []

                    for a in f["args"]:
                        arg = {}
                        arg["name"] = a["name"]
                        types = {}
                        types = recursive_oftype(f["type"])
                        arg["type"] = types

                        object["consumes"].append(arg)

                    query_list.append(object)
                    '''
            elif d["name"] == mutation_type_name:
                for f in d["fields"]:
                    object = {}
                    object["name"] = f["name"]
                    object["type"] = f["type"]
                    object["args"] = []

                    for a in f["args"]:
                        arg = {}
                        arg["name"] = a["name"]
                        object["args"].append(arg)

                    mutation_list.append(object)
            else:
                object = {}
                for f in d["fields"]:
                    object["name"] = f["name"]
                    object["type"] = f["type"]["kind"]
                object_list.append(object)
                '''
        elif d["kind"] == "SCALAR":
            scalar_list.append(d)
        elif d["kind"] == "ENUM":
            enum_list.append(d)

    return input_object_list, object_list, scalar_list, enum_list, query_list, mutation_list



data = parse(send_request())
pprint(data)


test = yaml.dump(data)
print(test)









